File: src/week4/P006.js
// 1. 최소 힙(Min Heap) 구현
// (JavaScript에는 표준 우선순위 큐가 없으므로 직접 구현)
class MinPriorityQueue {
    constructor() {
        this.heap = [];
    }

    isEmpty() {
        return this.heap.length === 0;
    }

    // 삽입 (O(log N))
    push(value) { // [거리, 노드]
        this.heap.push(value);
        this.bubbleUp();
    }

    // 삭제 (O(log N))
    pop() {
        if (this.isEmpty()) return null;
        if (this.heap.length === 1) return this.heap.pop();

        const min = this.heap[0];
        this.heap[0] = this.heap.pop();
        this.bubbleDown();
        return min;
    }

    // --- 내부 헬퍼 함수 ---
    bubbleUp() {
        let index = this.heap.length - 1;
        const lastNode = this.heap[index];

        while (index > 0) {
            const parentIndex = Math.floor((index - 1) / 2);
            const parentNode = this.heap[parentIndex];

            // 부모의 거리(cost)가 더 크면 swap
            if (parentNode[0] > lastNode[0]) {
                this.heap[index] = parentNode;
                index = parentIndex;
            } else {
                break;
            }
        }
        this.heap[index] = lastNode;
    }

    bubbleDown() {
        let index = 0;
        const rootNode = this.heap[index];
        const len = this.heap.length;

        while (true) {
            let leftChildIndex = index * 2 + 1;
            let rightChildIndex = index * 2 + 2;
            let swapIndex = -1;

            if (leftChildIndex < len) {
                if (this.heap[leftChildIndex][0] < rootNode[0]) {
                    swapIndex = leftChildIndex;
                }
            }
            
            if (rightChildIndex < len) {
                if (
                    (swapIndex === -1 && this.heap[rightChildIndex][0] < rootNode[0]) ||
                    (swapIndex !== -1 && this.heap[rightChildIndex][0] < this.heap[leftChildIndex][0])
                ) {
                    swapIndex = rightChildIndex;
                }
            }
            
            if (swapIndex === -1) break;

            this.heap[index] = this.heap[swapIndex];
            index = swapIndex;
        }
        this.heap[index] = rootNode;
    }
}

// 2. 입력 처리
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const [v, e] = input[0].split(' ').map(Number);
const snode = Number(input[1]);
const INF = 1e9; // 무한대 값

// 3. 그래프 및 최단 거리 테이블 초기화
const graph = Array.from({ length: v + 1 }, () => []);
const distance = new Array(v + 1).fill(INF);

// 4. 간선 정보 입력
for (let i = 2; i < e + 2; i++) {
    const [u, v, w] = input[i].split(' ').map(Number);
    // graph[u]에 [도착노드, 가중치] 저장
    graph[u].push([v, w]);
}

// 5. 다익스트라 함수
function dijkstra(start) {
    const pq = new MinPriorityQueue();

    // 1. 시작 노드 설정
    pq.push([0, start]); // [거리, 노드번호]
    distance[start] = 0;

    while (!pq.isEmpty()) {
        const [dist, now] = pq.pop();

        // 2. 이미 처리된 노드(더 짧은 경로를 찾은 경우)는 무시
        if (distance[now] < dist) {
            continue;
        }

        // 3. 현재 노드와 연결된 인접 노드 확인
        for (const [neighbor, weight] of graph[now]) {
            const cost = dist + weight; // 현재 노드를 거쳐 가는 비용

            // 4. 현재 노드를 거쳐 가는 것이 더 짧은 경우
            if (cost < distance[neighbor]) {
                distance[neighbor] = cost;
                pq.push([cost, neighbor]);
            }
        }
    }
}

// 6. 다익스트라 알고리즘 수행
dijkstra(snode);

// 7. 결과 출력
const output = [];
for (let i = 1; i <= v; i++) {
    if (distance[i] === INF) {
        output.push("INF");
    } else {
        output.push(distance[i]);
    }
}
console.log(output.join('\n'));

File: src/week4/P006.java
package week4;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class P006 {

    static class Node implements Comparable<Node> {
        int index; // 노드 번호
        int cost;  // 해당 노드까지의 거리 (가중치)

        public Node(int index, int cost) {
            this.index = index;
            this.cost = cost;
        }

        // 거리를 기준으로 오름차순 정렬
        @Override
        public int compareTo(Node other) {
            // int 비교는 단순 뺄셈으로 가능
            return this.cost - other.cost;
        }
    }

    static final int INF = (int) 1e9; // 무한대 값
    static ArrayList<ArrayList<Node>> graph;
    static int[] distance;
    static int v, e, snode; // v: 정점 수, e: 간선 수, snode: 시작 노드

    static void dijkstra(int start) {
        PriorityQueue<Node> pq = new PriorityQueue<>();

        // 1. 시작 노드 설정
        pq.add(new Node(start, 0));
        distance[start] = 0;

        while (!pq.isEmpty()) {
            Node node = pq.poll();
            int dist = node.cost; // 현재 노드까지의 거리
            int now = node.index; // 현재 노드 번호

            // 2. 이미 처리된 노드(더 짧은 경로를 찾은 경우)는 무시
            if (distance[now] < dist) {
                continue;
            }

            // 3. 현재 노드와 연결된 인접 노드 확인
            for (Node neighbor : graph.get(now)) {
                // neighbor.index = 인접 노드 번호
                // neighbor.cost = now에서 neighbor까지의 가중치
                int cost = dist + neighbor.cost; // 현재 노드를 거쳐 가는 비용

                // 4. 현재 노드를 거쳐 가는 것이 더 짧은 경우
                if (cost < distance[neighbor.index]) {
                    distance[neighbor.index] = cost;
                    // 우선순위 큐에 (갱신된 거리, 인접 노드) 추가
                    pq.add(new Node(neighbor.index, cost));
                }
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        v = Integer.parseInt(st.nextToken());
        e = Integer.parseInt(st.nextToken());
        snode = Integer.parseInt(br.readLine());

        // 1. 그래프 및 최단 거리 테이블 초기화
        graph = new ArrayList<>();
        for (int i = 0; i <= v; i++) {
            graph.add(new ArrayList<>());
        }
        distance = new int[v + 1];
        Arrays.fill(distance, INF);

        // 2. 간선 정보 입력
        for (int i = 0; i < e; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());      // 출발
            int v_node = Integer.parseInt(st.nextToken()); // 도착
            int w = Integer.parseInt(st.nextToken());      // 가중치
            // u번 노드에서 v_node번 노드로 가는 가중치 w
            graph.get(u).add(new Node(v_node, w));
        }

        // 3. 다익스트라 알고리즘 수행
        dijkstra(snode);

        // 4. 결과 출력
        StringBuilder sb = new StringBuilder();
        for (int i = 1; i <= v; i++) {
            if (distance[i] == INF) {
                sb.append("INF").append("\n");
            } else {
                sb.append(distance[i]).append("\n");
            }
        }
        System.out.print(sb.toString());
    }
}